instrumentation/pulsar/pulsar-2.8/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pulsar/v2_8/telemetry/PulsarSingletons.java

```diff
@@ -64,47 +64,61 @@ private static Instrumenter<PulsarRequest, Void> createConsumerReceiveInstrument
     MessagingAttributesGetter<PulsarBatchRequest, Void> getter =
         PulsarBatchMessagingAttributesGetter.INSTANCE;
 
-    return Instrumenter.<PulsarBatchRequest, Void>builder(
-            TELEMETRY,
-            INSTRUMENTATION_NAME,
-            MessagingSpanNameExtractor.create(getter, MessageOperation.RECEIVE))
-        .addAttributesExtractor(
-            createMessagingAttributesExtractor(getter, MessageOperation.RECEIVE))
-        .addAttributesExtractor(
-            NetClientAttributesExtractor.create(new PulsarNetClientAttributesGetter()))
-        .setEnabled(ExperimentalConfig.get().messagingReceiveInstrumentationEnabled())
-        .addSpanLinksExtractor(
-            new PulsarBatchRequestSpanLinksExtractor(
-                GlobalOpenTelemetry.getPropagators().getTextMapPropagator()))
-        .buildInstrumenter(SpanKindExtractor.alwaysConsumer());
+    boolean enabled = ExperimentalConfig.get().messagingReceiveInstrumentationEnabled();
+    InstrumenterBuilder<PulsarBatchRequest, Void> builder =
+        Instrumenter.<PulsarBatchRequest, Void>builder(
+                TELEMETRY,
+                INSTRUMENTATION_NAME,
+                MessagingSpanNameExtractor.create(getter, MessageOperation.RECEIVE))
+            .addAttributesExtractor(
+                createMessagingAttributesExtractor(getter, MessageOperation.RECEIVE))
+            .addAttributesExtractor(
+                NetClientAttributesExtractor.create(new PulsarNetClientAttributesGetter()))
+            .setEnabled(enabled);
+    if (enabled) {
+      builder.addSpanLinksExtractor(new PulsarBatchRequestSpanLinksExtractor(PROPAGATOR));
+    }
+    return builder.buildInstrumenter(SpanKindExtractor.alwaysConsumer());
   }
 
   private static Instrumenter<PulsarRequest, Void> createConsumerProcessInstrumenter() {
     MessagingAttributesGetter<PulsarRequest, Void> getter =
         PulsarMessagingAttributesGetter.INSTANCE;
 
-    return Instrumenter.<PulsarRequest, Void>builder(
-            TELEMETRY,
-            INSTRUMENTATION_NAME,
-            MessagingSpanNameExtractor.create(getter, MessageOperation.PROCESS))
-        .addAttributesExtractor(
-            createMessagingAttributesExtractor(getter, MessageOperation.PROCESS))
-        .buildInstrumenter();
+    boolean enabled = ExperimentalConfig.get().messagingReceiveInstrumentationEnabled();
+    InstrumenterBuilder<PulsarRequest, Void> builder =
+        Instrumenter.<PulsarRequest, Void>builder(
+                TELEMETRY,
+                INSTRUMENTATION_NAME,
+                MessagingSpanNameExtractor.create(getter, MessageOperation.PROCESS))
+            .addAttributesExtractor(
+                createMessagingAttributesExtractor(getter, MessageOperation.PROCESS));
+    if (!enabled) {
+      builder.addSpanLinksExtractor(new PulsarRequestSpanLinksExtractor(PROPAGATOR));
+    }
+
+    return builder.buildInstrumenter();
```

Comment:
If the receive telemetry is disabled I believe the process span should use the context extracted from the incoming message as the parent -- at least that's how the other instrumentations are behaving.

Also it should be a consumer span (`builder.buildConsumerInstrumenter()`), not an internal span.

Reply:
I think a process span should be a internal span, because if `otel.instrumentation.messaging.experimental.receive-telemetry.enabled` set to true, there will be 2 consumer span. 
So in response to this situation, I believe that internal span is better, WDYT?

Reply:
The consensus we arrived at is to use `CONSUMER` for both, see example in the spec: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#batch-receiving

Reply:
hi @dao-jun! just checking in if you were able to update these to follow the messaging semantic conventions?

Reply:
@mateuszrzeszutek @trask fixed, sorry for delayed such long time